torch/distributed/_composable/replicate.py

```diff
@@ -0,0 +1,96 @@
+import torch
+import torch.nn as nn
+from . import _ddp
+
+from typing import List, Tuple
+
+
+class DistributedState:
+    ...
+
+
+class ReplicateState(DistributedState):
+    def __init__(self) -> None:
+        self.modules: List[nn.Module] = []
+        self.has_initialized: bool = False
+        self._param_list: nn.ParameterList = nn.ParameterList()
+
+    def mark_modules(self, *modules: nn.Module) -> None:
+        for module in modules:
+            self.modules.append(module)
+            # TODO(@yhcharles): do not directly set attribute
+            module._distributed_state = self  # type: ignore[assignment]
+
+    def _recursive_collect_params(self, module: nn.Module) -> None:
+        if (
+            getattr(module, "_distributed_state", None) is not None
+        ):  # managed by another API
+            return
+
+        self._param_list.extend(
+            param for param in module.parameters() if param.requires_grad
+        )
```

Comment:
No, it's not. It's top-down search. The assumption is: whenever user marks a module, then this module and all its sub-modules (as long as them are not managed by other APIs) will be managed by `replicate()`. Does this sound good to you?

Reply:
same here, if it is top-down search, the implementation seems to collect duplicate parameters?

Reply:
I'm not sure I get your point. Why do you think it collects duplicate parameters?
My idea is when it runs recursively on all sub-modules, it will first check if the sub-module is already marked by another API, if true, it will skip that sub-tree, so there won't be any duplicate?

Reply:
module.parameters() will include children's parameters, for the case like this, b and d are marked to be replicated, e is marked to be sharded, what will 'self._param_list' be?

              root
            /     |     \
          a      b     c
                /    \
               d    e

Reply:
I guess the best way is to write some unit tests to verify the parameters are collected as expected